Add popover support to select control dropdown

louwie17 opened the pull request:
### All Submissions:

-   [x] Have you followed the [WooCommerce Contributing guideline](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md)?
-   [x] Does your code follow the [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/)?
-   [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change?

<!-- Mark completed items with an [x] -->

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

### Changes proposed in this Pull Request:

This adds support for the SelectControl dropdown to be rendered in a `Popover.Slot` component, this is helpful for when we display dropdown's within Modal's.

<img width="908" alt="Screen Shot 2022-10-05 at 4 40 16 PM" src="https://user-images.githubusercontent.com/2240960/194148070-6e1a69b8-c6fd-4e44-a988-54859248f3b0.png">

### How to test the changes in this Pull Request:

1. Load this branch, run `pnpm run build && pnpm --filter=@woocommerce/storybook storybook`
2. Go to the experiment Select control and select the `Single Within Modal Using Body Dropdown Placement` example
3. It should show a modal, the dropdown should work as expected and not be hidden by the modal, overlap the Modal container in this
4. Test the other SelectControl stories, they should still work as expected.

### Other information:

-   [x] Have you added an explanation of what your changes do and why you'd like us to include them?
-   [ ] Have you written new tests for your changes, as applicable?
-   [ ] Have you successfully run tests with your changes locally?
-   [x] Have you created a changelog file for each project being changed, ie `pnpm --filter=<project> run changelog add`?

<!-- Mark completed items with an [x] -->

### FOR PR REVIEWER ONLY:

-   [ ] I have reviewed that everything is sanitized/escaped appropriately for any SQL or XSS injection possibilities. I made sure Linting is not ignored or disabled.

louwie17 committed:
Add dropdownPlacement option to SelectControl

louwie17 committed:
Make use of parent element of menu for width

louwie17 committed:
Add changelog

louwie17 commented on packages/js/components/src/experimental-select-control/select-control.tsx:
```diff
@@ -42,6 +42,7 @@ type SelectControlProps< ItemType > = {
 	onSelect?: ( selected: ItemType ) => void;
 	placeholder?: string;
 	selected: ItemType | ItemType[] | null;
+	dropdownPlacement?: 'inline' | 'body';
```
Using `body` will depend on the `Popover.Slot` being used, maybe I should add a warning if that is not available?
If `Popover.Slot` is not used it will just default to rendering it `inline`.

louwie17 committed:
Add useEffect for boundingRect so it gets updated accordingly

mdperez86 commented on packages/js/components/src/experimental-select-control/menu.tsx:
```diff
@@ -29,7 +70,7 @@ export const Menu = ( { children, getMenuProps, isOpen }: MenuProps ) => {
 			) }
 		>
-			{ isOpen && children }
+			{ isOpen && childrenPopoverMarkup }
```
This is semantically incorrect. An `ul` element must have `li` elements as its direct children. See `Permitted content` section from [MDN](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/ul). The `ul` element should be at line 44 replacing the `div` and `children` must be `li` elements.

louwie17 replied:
Thanks, yeah I just fixed this as part of: https://github.com/woocommerce/woocommerce/pull/34967/commits/043975f49ca8e259a91e9bcfc4adcdeaec63cf58

louwie17 replied:
It was causing some styling glitches to, as the screenshot above shows well.

louwie17 committed:
Fix styling for popover dropdown

louwie17 commented:
@mdperez86 this should be good for a re-review

joshuatf commented on packages/js/components/src/experimental-select-control/menu.tsx:
```diff
@@ -16,9 +17,48 @@ type MenuProps = {
+	}, [ selectControlMenuRef.current ] );
+
+	/* eslint-disable jsx-a11y/no-noninteractive-element-interactions, jsx-a11y/click-events-have-key-events */
+	const childrenPopoverMarkup = (
```
Having component markup this large inside another component feels a bit like a code smell.  Should we move it inside the render below or to a separate popover component file?

louwie17 replied:
Thanks Josh I moved it down to the render return, as pushing it to a different component felled a bit like over kill as it practically required all the same `props`: https://github.com/woocommerce/woocommerce/pull/34967/commits/8b4c672c2615805a1caf1a5a84017a3593232bbe

joshuatf reviewed (dismissed):
This is testing well for me!  Double check with a screen reader and didn't notice any regressions.

Left one comment about where the popover should live, but pre-approving ✅

louwie17 committed:
Move popup markup down to render return

louwie17 commented:
@joshuatf & @mdperez86 thanks for the reviews, would either of you be able to re-review this, I just addressed Josh's suggestions and moved the mark up down.

joshuatf reviewed (dismissed):
Thanks for moving that markup, @louwie17!  LGTM 🚢

louwie17 committed:
Move getMenuProps up

louwie17 commented:
@joshuatf this should be good for a re-review, but probably needs to be re-tested with a screen reader if you can.

joshuatf commented on packages/js/components/src/experimental-select-control/menu.tsx:
```diff
@@ -16,9 +17,24 @@ type MenuProps = {
-		<ul
-			{ ...getMenuProps() }
+		<div
+			{ ...getMenuProps( { ref: selectControlMenuRef } ) }
```
This breaks accessibility and moves the `role="listbox"` to the div instead of the ul.  Items will not be read aloud in screen readers.

Is there a reason we need to move the menu props to the div?

louwie17 replied:
I was afraid of that, when it was on the other element `downshift` showed an error that the `ref` was not used correctly. My guess is because of it being inside of the Popover and thus rendered at a later time, I can see if I can change that, or leave the `getMenuProps` on the nested element but pass in the ref of the parent element (as I am not sure what the `ref` will exactly be used for.

louwie17 replied:
@joshuatf I guess the main issue was that I was conditionally rendering the `popover` previously. Apparently as part of accessibility ( [see comment](https://github.com/downshift-js/downshift/issues/1272#issuecomment-1236290430)) the menu should always be rendered.
I moved the `getMenuProps` back down and removed the `isOpen` condition and moved that down to the children (as that is fine to be conditionally rendered).
This should work as expected again.

louwie17 replied:
This is the commit btw: https://github.com/woocommerce/woocommerce/pull/34967/commits/a12b0c066cab1cff9063b8e2185ba7439a668020

joshuatf reviewed (dismissed):
The latest changes breaks accessibility.  Is there a reason we needed to move those props?

louwie17 committed:
Move getMenuProps down to list again and always render the popover menu

mattsherman commented on packages/js/components/src/experimental-select-control/menu.tsx:
```diff
@@ -16,9 +17,24 @@ type MenuProps = {
+		}
+	}, [ selectControlMenuRef.current ] );
+
+	/* eslint-disable jsx-a11y/no-noninteractive-element-interactions, jsx-a11y/click-events-have-key-events */
```
I assume these eslint disables are because of the `onMouseUp` below. It would be nice to have a comment right here saying why these checks are being disabled though.

mattsherman reviewed (changes requested):
In general, this works as expected, when not considering accessibility.

One issue I noticed is that if you hit the escape key when the select control has focus, it dismisses the entire modal, not just the select control's dropdown.

There are definitely some accessibility issues, as this video demonstrates (I took the low-tech way and recorded my screen with my iPhone... didn't want to mess with a loopback audio adapter on my Mac, due to potential security concerns)...

https://user-images.githubusercontent.com/2098816/195916923-3569269a-5c5f-451f-8e5b-bda47ac0f6ef.mov

Issues:

* Often, when typing, it starts to announce the number of results and then cuts itself off and re-announces what was just typed
* There is no announcement of the item you are selecting when you arrow through the results in the popover.

louwie17 committed:
Add MenuSlot that adds popover slot to mody with aria-live=off so it is skipped by Modal

louwie17 committed:
Add comment in relation to the eslint-disable comment

louwie17 commented:
@mattsherman thanks for the review on this Matt, I have updated the PR and addressed the accessibility concerns, see comments on each point below:

> * Often, when typing, it starts to announce the number of results and then cuts itself off and re-announces what was just typed

I am not sure what the case is here, but after my above changes this only seemed to happen after the first character, but not after any remaining characters. As mentioned in Slack I didn't notice this within the attribute modal in the product MVP but only in storybook, so I am not sure if it could be related to that.

> * There is no announcement of the item you are selecting when you arrow through the results in the popover.

This should be fixed now, turns out that the modal component has some logic that add's `aria-hidden` to all the root elements that isn't the modal, or do not have the `aria-live` tag on them. See the code here: https://github.com/WordPress/gutenberg/blob/trunk/packages/components/src/modal/aria-helper.ts. This in turn caused the popover slot (containing the dropdown) to be hidden and thus never get announced.

In order to fix this I had to add the Popover.Slot to the root element and wrap it in a `div` with `aria-live=off` so it would be skipped by the Modal logic.
Given this is not something we want our consumers to worry about, I added a `MenuSlot` component that does this, see the bottom of the `Menu.tsx` file.

mattsherman commented:
> > * Often, when typing, it starts to announce the number of results and then cuts itself off and re-announces what was just typed
> 
> I am not sure what the case is here, but after my above changes this only seemed to happen after the first character, but not after any remaining characters. As mentioned in Slack I didn't notice this within the attribute modal in the product MVP but only in storybook, so I am not sure if it could be related to that.

Odd. I suppose if it only happens in Storybook we can let it slide for now.

> > * There is no announcement of the item you are selecting when you arrow through the results in the popover.
> 
> This should be fixed now, 

I can confirm... it now announces the selected item when arrowing through!

mattsherman commented:
@louwie17 

A few issues I've noticed...

- You can't double-click text in the select control to select it (not sure if this is new to this PR or it is existed before)
- The dropdown menu flashes in the upper left of the screen. This seems to happen when you've typed in text that doesn't match anything (causing the dropdown to disappear) and then change text until there are matches again (causing it to re-appear)

https://user-images.githubusercontent.com/2098816/196479457-6c13af04-6c42-4e9f-950f-6e3ff45057cf.mov

mattsherman commented on packages/js/components/src/experimental-select-control/stories/index.tsx:
```diff
@@ -1,17 +1,24 @@
+} from '@wordpress/components';
 import React from 'react';
-import { createElement, useState } from '@wordpress/element';
+import { useState, createElement } from '@wordpress/element';
```
Why did the order of these change? I thought that our guidelines were to be alphabetical. Not a big deal, honestly.

mattsherman commented on packages/js/components/src/experimental-select-control/menu.tsx:
```diff
@@ -22,9 +29,25 @@ export const Menu = ( {
+
+	useEffect( () => {
+		if (
+			selectControlMenuRef.current &&
```
You can get rid of this, right? the `selectControlMenuRef.current?.parentElement` accounts for this with the conditional chaining.

louwie17 replied:
Done as part of https://github.com/woocommerce/woocommerce/pull/34967/commits/8e895d4139b8369fd8c90943818bb20c7d10cf06

mattsherman commented on packages/js/components/src/experimental-select-control/menu.tsx:
```diff
@@ -22,9 +29,25 @@ export const Menu = ( {
+			selectControlMenuRef.current?.parentElement
+		) {
+			setBoundingRect(
+				selectControlMenuRef.current?.parentElement.getBoundingClientRect()
```
you don't need the `?` here either.

louwie17 replied:
Done as part of https://github.com/woocommerce/woocommerce/pull/34967/commits/8e895d4139b8369fd8c90943818bb20c7d10cf06

mattsherman commented on packages/js/components/src/experimental-select-control/menu.tsx:
```diff
@@ -36,7 +59,43 @@ export const Menu = ( {
+					'woocommerce-experimental-select-control__popover-menu',
+					{
+						'is-open': isOpen,
+						'has-results': Array.isArray( children )
```
I think `'has-results': Array.isArray( children ) && children.length > 0` be a bit clearer. Seems a bit less clear to be setting `has-results` to `children.length` or `Boolean (children)`.

louwie17 replied:
Done as part of https://github.com/woocommerce/woocommerce/pull/34967/commits/8e895d4139b8369fd8c90943818bb20c7d10cf06

mattsherman reviewed (changes requested):
The flashing of the dropdown is the only real blocker for me. Left a few minor code readability comments as well.

louwie17 committed:
Fix flashing of popover and addressed some minor PR suggestions

louwie17 commented:
> The flashing of the dropdown is the only real blocker for me. Left a few minor code readability comments as well.

Thanks, I fixed this, the issue was around the `display:none` on the original parent element that I had left in. This caused the popover to re-calculate the position everytime the popover was closed or opened where it would default to top: 0; left: 0; causing the weird jump.
Removing that allowed for the top, left values to not change, so no more jumping behaviour :) 

> This still happens. Btw, it only happens when VoiceOver is not enabled -- I imagine VoiceOver overrides the ESC key?

Strange it doesn't happen when VoiceOver is not enabled, this does seem to be a bug with either Downshift or Modal, It does appear that Downshift triggers the `event.preventDefault()` -> https://github.com/downshift-js/downshift/blob/master/src/hooks/useCombobox/index.js#L201
Which the `Modal` is suppose to hold into account but it seemingly isn't -> https://github.com/WordPress/gutenberg/blob/trunk/packages/components/src/modal/index.tsx#L106

Something to be updated in a future PR I think.

> You can't double-click text in the select control to select it (not sure if this is new to this PR or it is existed before)

Hmm yeah that is a Select Control bug I think, I tested it on trunk and it's the same behaviour, I can create a separate issue for this.

mattsherman reviewed (dismissed):
Thanks for the final changes, @louwie17 ... works as expected. No more flash. Nice job!

mattsherman reviewed (approved):
Still looks good after rebase. :shipit:
